### ISSUE TITLE
gdb.debug() over ssh - gdb trying to use remote exe path on local host and not finding file?

### ISSUE THREAD

**Issue by ghost** [Gallopsled/pwntools#1117]
I'm trying to use gdb.debug() and ssh tubes to run an executable on a remote host and then debug it with gdb.  As shown in the ssh [gdb.debug()](http://docs.pwntools.com/en/stable/gdb.html#pwnlib.gdb.debug) example in the docs.

When I get a new terminal with the gdb session there is an error saying the file (as specified in the call to gdb.debug() ) is not found, and the error lists the path to the executable on the remote host.  It looks as though gdb is trying to use location of the executable on the remote machine's file system..but on the local machine. 

Below I've included

- The basic pwntools Python script
- The execution output of the script
- The gdb output from the new terminal
- Directory listing on the remote host showing the location of the executable I'm trying to debug

Environment:  
- Local: Ubuntu 17.10 x86-64, gdb 8.0.1
- Remote host: Ubuntu 14.04.5 x86-32, gdbserver 7.7.1 Local KVM VM at IP of 192.168.122.55.
- pwntools version: 3.12.0
- The name 'mbe_testlab' (used for the ssh connection hostname in the script) is a Host defined in my ssh config file. 

**ssh_gdb_example.py - copied from pwntools docs for gdb.debug() & modified for ssh details, executable**
```
#!/usr/bin/env python

from pwn import *

#host running script is x86-64, ssh remote host is x86-32
context.arch='i386'
context.bits=32

# Connect to the SSH server
shell = ssh('mbe_test','mbe_testlab', keyfile='~/.ssh/local_vms')

# Start a process on the server
io = gdb.debug(['/home/mbe_test/mbe_dev/bin/lab3C'], exe='/home/mbe_test/mbe_dev/bin/lab3C',
                               ssh=shell,
                               gdbscript='''
break main
continue
''')


# Interact with the process
io.interactive()
```

**Execution output of script**

```
pgrohe@zen:~/projects/pwntools_learning$ python ssh_gdb_example.py
[+] Connecting to 192.168.122.55 on port 22: Done
[!] Couldn't check security settings on '192.168.122.55'
[+] Opening new channel: 'export PATH=$PATH:$PWD; which gdbserver': Done
[+] Receiving all data: Done (19B)
[*] Closed SSH channel with 192.168.122.55
[+] Starting remote process '/usr/bin/gdbserver' on 192.168.122.55: pid 2581
[+] Connecting to 127.0.0.1:43936 via SSH to 192.168.122.55: Done
[+] Trying to bind to 0.0.0.0 on port 0: Done
[←] Waiting for connections on 0.0.0.0:33779
[+] Downloading '/home/mbe_test/mbe_dev/bin/lab3C' to 'lab3C': Found '/home/mbe_test/mbe_dev/bin/lab3C' in ssh cache
[*] running in new terminal: /usr/bin/gdb -q  "lab3C" -x "/tmp/pwnBowLVC.gdb"
[*] Switching to interactive mode
$ 
```

**Debug terminal spawned**

```
pwndbg: loaded 164 commands. Type pwndbg [filter] for a list.
pwndbg: created $rebase, $ida gdb functions (can be used with print/break)
Set Source code / assembly syntax highlight stylename of pygments module to 'pastie'
Set color for flags register (flag changed) to 'bold,red'
Set color for flags register (flag set) to 'bold'
Set color for flags register (flag unset) to 'normal'
Set color for hexdump command (special bytes) to 'purple'
Set color added to highlights like source/pc to 'bold'
Set color for stack memory to 'green'
Set color of breakpoint messages to 'red,bold'
Set color of hint and marker messages to 'bold'
Set color of warning messages to 'bold'
Set Source code / assembly syntax highlight stylename of pygments module to 'pastie'
Set whether to clear the screen before printing the context to True
Set number of additional lines to print for the nearpc command to 16
Set whether to show flags registers to True
Reading symbols from lab3C...(no debugging symbols found)...done.
/tmp/pwnBowLVC.gdb:2: Error in sourced command file:
/home/mbe_test/mbe_dev/bin/lab3C: No such file or directory.
pwndbg> 

```

**Remote host - Confirmation /home/mbe_test/mbe_dev/bin/lab3C exists**
```
mbe_test@mbe-testlab:~/mbe_dev/bin$ pwd
/home/mbe_test/mbe_dev/bin
mbe_test@mbe-testlab:~/mbe_dev/bin$ ls -alh
total 16K
drwxrwxr-x 2 mbe_test mbe_test 4.0K Feb 18 17:01 .
drwxrwxr-x 7 mbe_test mbe_test 4.0K Feb 19 00:19 ..
-rwxrwxr-x 1 mbe_test mbe_test 7.4K Feb 18 17:01 lab3C
mbe_test@mbe-testlab:~/mbe_dev/bin$ file lab3C 
lab3C: ELF 32-bit LSB  executable, Intel 80386, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.24, BuildID[sha1]=9fe71094eafd74bd3d7482bd4e6c000e67159e95, not stripped
```

Let me know if you need any more info.  Cheers.

**Comment by ghost** [comment]
With PWNLIB_DEBUG=1 the listed contents of the .gdb file to execute via '-x' is

```
[DEBUG] Wrote gdb script to '/tmp/pwnHLl5Ci.gdb'
    file /home/mbe_test/mbe_dev/bin/lab3C
    target remote 127.0.0.1:34513
    file /home/mbe_test/mbe_dev/bin/lab3C
    
    break main
    continue
```

Those GDB commands as shown depend on having the same directory hierarchy on both the local machine & the remote host where the executable is run no?   

Wouldn't we need both:
- `set remote exec-file <path to executable on remote host>`
- `file <path to local copy of executable>`

**Comment by njfox** [comment]
I'm having this same issue in a fresh venv on Arch Linux (also MBE, lab 6B).

**Comment by gymgit** [comment]
I have been having the same problem and started looking into it, the issue is in `pwnlib/gdb.py:attach()` the supplied `exe` function parameter is both used as a local and remote path. If these path don't exist on both systems the error occurs. 

* line 726: the `exe` param is appended to the gdb script as a local path. Note that the `exe` param has already been used similarly in `pwnlib/gdb.py:debug()` if that is the caller.

```python
    if exe:
        # The 'file' statement should go first
        pre = 'file %s\n%s' % (exe, pre) 
```

* line 742: the `exe` parameter is used to download the target file from the remote ssh server.

```python
    if exe and context.native:
        if ssh:
            ssh.download_file(exe)
            exe = os.path.basename(exe)
        if not os.path.isfile(exe):
            log.error('No such file: %s' % exe)
        cmd += ' "%s"' % exe
```

I think the correct solution is what @pgrohe mentioned, we need two different parameters, the current `exe` and a separate `remote_exe`. If `remote_exe` is specified it can be used to download the target file to the `exe` path.

I will try to create a pr later.

**Comment by zachriggle** [comment]
`exe`  has always referred to a path on-disk on the host that will execute the binary.

For local debugging, this is a local path.

For remote debugging, this is a remote path.

It does appear that `attach(..., ssh=...)` hits this code path when invoked directly, instead of via `gdb.debug()` (which does port forwarding for gdbserver, instead of spawning a remote GDB instance).

**Comment by zachriggle** [comment]
Check out #1230 as an alternative to #1229.  I haven't tested it, but I believe this should resolve the issue.

**Comment by gymgit** [comment]
I think there is a confusion here how `exe` is used when debugging remotely. You mention that it is meant to be used as a remote path, but both in `gdb.debug()` and in `gdb.attach()` it is added to the gdb script (that it is used locally) as a `file` parameter. This needs to be a local path as gdb tries
 to read this file on the local system.

Initially, I experienced this issue using the `gdb.debug()` function, I never tried it with calling `attach(..., ssh=...)` directly, but it should provide the same result.

#1230 fixes the issue if `exe` is used as a local path then the gdb script would be generated correctly.

**Comment by Arusekk** [comment]
This seems resolved by #1230
